Java: model top JDK APIs

jcogs33 opened the pull request:
### Description:
This PR adds missing MaD models for the next 100 top JDK APIs. (The first 100 were reviewed and modeled in https://github.com/github/codeql/pull/11572).

### Consideration:
Please review the added models in detail, especially the following: 
 - Why does `java.lang.String#endsWith` have an existing summary model instead of a neutral model? It returns a boolean, so I thought it should have a neutral model. Should I leave this one as a summary model or switch it to a neutral model?
 - Does the model for `java.lang.System#getProperty(String)` look correct? I couldn't seem to get a working test case for this one.
 - Should any of the following be summary models instead of neutrals?
     - `java.lang.Class#getClassLoader()`
     - `java.text.DateFormat#format(Date)`
     - `java.text.SimpleDateFormat#SimpleDateFormat(String)`
     - `java.time.LocalDate#of(int,int,int)`
     - `java.util.function.Consumer#accept(Object)`
     - `java.util.Date#Date(long)`
     - `java.util.stream.Collectors#joining(CharSequence)`
     - `java.util.stream.Collectors#toMap(Function,Function)`

egregius313 commented on java/ql/lib/ext/java.lang.model.yml:
```diff
@@ -37,22 +37,32 @@ extensions:
       - ["java.lang", "String", False, "concat", "(String)", "", "Argument[0]", "ReturnValue", "taint", "manual"]
       - ["java.lang", "String", False, "copyValueOf", "", "", "Argument[0]", "ReturnValue", "taint", "manual"]
-      - ["java.lang", "String", False, "endsWith", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
+      - ["java.lang", "String", False, "endsWith", "", "", "Argument[-1]", "ReturnValue", "taint", "manual"] # ! why is this a summary model and not a neutral model instead? it returns a boolean
```
I would also like to know if this should be moved to extending `neutralModel`.

michaelnebel replied:
I think the "endsWith" should be a neutral model. Even. though taint is propagated the `boolean` type can be considered a form of sanitizer - this is what we do for model generation.

atorralba replied:
See my comment below, but for booleans I feel slightly inclined to consider them neutral models as well (contrary to numbers).

jcogs33 replied:
@atorralba or @michaelnebel Switching `endsWith` to a neutral model is causing failures for two test cases: `semmle-code/semmlecode-tests/external-tests/DO_NOT_DISTRIBUTE/security/CWE-807/CWE/examples/ConditionalBypass.qlref` and `semmle-code/ql/java/ql/test/library-tests/dataflow/taint/test.ql`. 
Should I go ahead and update these test cases? Or not switch `endsWith` to neutral after all?

jcogs33 replied:
Update in https://github.com/github/codeql/pull/11779/commits/ffb267937acb5585c69460d86eb54d13e177df21, https://github.com/github/codeql/pull/11779/commits/a39b2aaaac8f9904979f9e9b8a582ba3290916d6, and https://github.com/github/codeql/pull/11779/commits/c3a1d088ace1905fd9b34b23de80300530683989.

egregius313 reviewed (commented):
LGTM. I can approve once we get an answer about the `endsWith` row.

jcogs33 commented:
> Unlike for #11572, there seem to be quite a few alert changes on DCA for this PR. Has this already been spot-checked for anything unexpected?

Not yet. I was planning to do another DCA run after addressing review comments (since I assumed that I would likely need to make changes to some models), and I'll review that DCA run in more detail.

michaelnebel commented on java/ql/lib/ext/java.lang.model.yml:
```diff
@@ -88,33 +99,45 @@ extensions:
       - ["java.lang", "StringBuffer", True, "StringBuffer", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]
       - ["java.lang", "StringBuilder", True, "StringBuilder", "", "", "Argument[0]", "Argument[-1]", "taint", "manual"]
       - ["java.lang", "System", False, "arraycopy", "", "", "Argument[0]", "Argument[2]", "taint", "manual"]
+      - ["java.lang", "System", False, "getProperty", "(String)", "", "Argument[-1].MapValue", "ReturnValue", "value", "manual"]
```
Isn't `getProperty` a static method in the `System` class?
I would have expected `getProperty : Argument[0] -> ReturnValue`

jcogs33 replied:
> Isn't `getProperty` a static method in the `System` class?

yes

> I would have expected `getProperty : Argument[0] -> ReturnValue`

I might be overthinking this, but `Argument[0] -> ReturnValue` didn't quite make sense to me for this one since `Argument[0]` is a key, and the `ReturnValue` should be the value that the key maps to (based on my understanding of the [docs](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getProperty-java.lang.String-)), so I was trying to take that mapping into account. Is taking the mapping into account not necessary for this model?

atorralba replied:
Yes, propagating the taint from the argument to the return value isn't correct because controlling the key doesn't imply controlling the value stored with that key in the system properties.

Since a static method doesn't really have a useful qualifier, I think you'll need to use [synthetic globals](https://github.com/github/codeql/pull/10814/) for this. See the test for an example with arrays (would need to be adapted to maps):

```ql
"my.qltest.synth;A;false;storeInArray;(String);;Argument[0];SyntheticGlobal[db1].ArrayElement;value;manual",
"my.qltest.synth;A;false;readArray;();;SyntheticGlobal[db1].ArrayElement;ReturnValue;value;manual",
```

Now if you only want to create edges when the key used to set the property matches the key used to read the property, then I think MaD isn't enough and you'll need to use regular CodeQL (probably an `AdditionalValueStep`, or synthetic callables + synthetic globals for more advanced cases). Which is (probably?) out of the scope of this PR.

jcogs33 replied:
Thanks! Updated in https://github.com/github/codeql/pull/11779/commits/0c7ffb055471c116ed26e0e244b3b517aa61d4ce. Let me know if it does not look correct.

> Which is (probably?) out of the scope of this PR.

I agree that that sounds out of scope for this PR.

atorralba replied:
We should keep an eye on this in DCA, it may cause a lot of FP flow from a `set` of a property to a `get` of a different one.

atorralba replied:
See https://github.com/github/codeql/pull/11779#discussion_r1071233634

michaelnebel commented on java/ql/lib/ext/java.math.model.yml:
```diff
@@ -4,3 +4,11 @@ extensions:
       extensible: summaryModel
     data:
       - ["java.math", "BigDecimal", False, "BigDecimal", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]
+      - ["java.math", "BigDecimal", False, "valueOf", "(double)", "", "Argument[0]", "ReturnValue", "taint", "manual"]
```
neutral

michaelnebel commented on java/ql/lib/ext/java.math.model.yml:
```diff
@@ -4,3 +4,11 @@ extensions:
     data:
       - ["java.math", "BigDecimal", False, "BigDecimal", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]
+      - ["java.math", "BigDecimal", False, "valueOf", "(double)", "", "Argument[0]", "ReturnValue", "taint", "manual"]
+      - ["java.math", "BigDecimal", False, "valueOf", "(long)", "", "Argument[0]", "ReturnValue", "taint", "manual"]
```
neutral

michaelnebel commented on java/ql/lib/ext/java.sql.model.yml:
```diff
@@ -19,4 +19,13 @@ extensions:
       pack: codeql/java-all
       extensible: summaryModel
     data:
+      - ["java.sql", "PreparedStatement", True, "setInt", "(int,int)", "", "Argument[1]", "Argument[-1]", "value", "manual"]
```
neutral

michaelnebel commented on java/ql/lib/ext/java.sql.model.yml:
```diff
@@ -19,4 +19,13 @@ extensions:
     data:
+      - ["java.sql", "PreparedStatement", True, "setInt", "(int,int)", "", "Argument[1]", "Argument[-1]", "value", "manual"]
+      - ["java.sql", "PreparedStatement", True, "setString", "(int,String)", "", "Argument[1]", "Argument[-1]", "value", "manual"]
+      - ["java.sql", "ResultSet", True, "getInt", "(String)", "", "Argument[-1]", "ReturnValue", "taint", "manual"]
```
neutral

michaelnebel commented:
@atorralba : What is your view on simple types as sanitizers?
Assume we have functions like:
f1 : int -> string
f2 : string -> int
f3 : string -> bool
then even if there is taint flow, shouldn't we then consider these to be neutrals because we use primitive types as intermediates (in this case either as the source or return type)?

atorralba commented:
> @atorralba : What is your view on simple types as sanitizers?

I agree that normally primitive types aren't likely to carry relevant taint. Nonetheless, there're a few queries that are interested in tainted primitives (e.g. a DoS by sleep, in which a tainted integer reaches a sleep function — an attacker could specify a high value and cause the program to sleep for a long time, practically causing a DoS). That's why in the previous round of top JDK APIs modelling, I recommended @jcogs33 to consider models that propagated taint through numbers (ints, longs, doubles...). Booleans and chars are different, but there's _probably_ a query out there that could be interested in them as well.

So the approach in my mind was adding these models, and then queries not interested in primitive types (the majority of them) would need to add a primitive type sanitizer. Most injection queries already do. There were discussions with @aschackmull about generalizing this kind of sanitizer so that we could reuse it more easily.

Now, if the model generator is excluding this kind of models by default, then I'm not sure if we should add them manually. Ideally the criteria for both generating and manually creating models should be the same. We could either modify the generator, or change our current criteria: instead of adding default taint steps for primitives types, queries interested in them should add additional taint steps specifically in their configurations. This could also be generalized as reusable taint steps, probably.

I'm open to either approach, so I'd be happy to hear more opinions.

joefarebrother commented on java/ql/lib/ext/java.util.stream.model.yml:
```diff
@@ -92,4 +92,7 @@ extensions:
       pack: codeql/java-all
       extensible: neutralModel
     data:
+      - ["java.util.stream", "Collectors", "joining", "(CharSequence)", "manual"]
```
This seems like it ought to be taint preserving, though I'm not sure how it would fit in with the special-case handling of the collect api

jcogs33 replied:
I've added a summary model for this for flow from `Argument[0] -> ReturnValue` in  https://github.com/github/codeql/pull/11779/commits/181a711f045012f91f06565e77ba43a633304f97. It does feel somewhat incomplete without having the `collect` api handled by MaD as well though 🤔

atorralba replied:
`Collector` APIs are covered [here](https://github.com/github/codeql/blob/ce06df315236eb011ffa269ef5a88e2527af4424/java/ql/lib/semmle/code/java/frameworks/Stream.qll#L44). As you said, this model isn't going to do much without another model that reads taint from the collector, which doesn't currently exist. I _think_ it won't hurt having it either, so we could keep it — with the risk of causing confusion to a future reader that wonders why we modeled this incompletely in MaD.

jcogs33 replied:
> So it's probably best to remove it.

I've removed it in https://github.com/github/codeql/pull/11779/commits/ac064ac2a78aae7c7096a519949e30765f3c58ad

yo-h commented:
> What is your view on simple types as sanitizers?

I wonder whether we should add additional kinds of summaries - perhaps in addition to "value" and "taint", we may want "value-numeric" and "taint-numeric" or similar. Then, those models would be excluded by default for most queries (neutral) but could be enabled for queries where this kind of flow is relevant.

That said, the above is out of scope for this PR. As a more immediate resolution, we could add numeric flow as neutral, along with a comment that marks it as numeric flow. These comments would serve as a reminder for later generalizations of the dataflow. (cc @aschackmull)

atorralba commented:
> As a more immediate resolution, we could add numeric flow as neutral, along with a comment that marks it as numeric flow.

👍 We'll then probably need to do the same for numeric flow models added in the previous PR, for consistency.

atorralba commented:
> Should any of the following be summary models instead of neutrals?
>
> `java.lang.Class#getClassLoader()`

No, controlling the class (which is in itself ambiguous and unlikely) doesn't imply controlling its class loader.

> `java.text.DateFormat#format(Date)`

I think `Date` is a restricted-enough type to be in the same category as primitive types, in the sense that it can hardly carry significant taint. So, following [the suggestion above](https://github.com/github/codeql/pull/11779#issuecomment-1378207541), I wouldn't add a model for this.

> `java.text.SimpleDateFormat#SimpleDateFormat(String)`

Same reasoning: a `SimpleDateFormat` will hardly be tainted in a relevant way.

> `java.time.LocalDate#of(int,int,int)`

Same.

> `java.util.function.Consumer#accept(Object)`

This is related to specialized lambda flow, so I don't think it should be handled by MaD.

> `java.util.Date#Date(long)`

Same as above.

`java.util.stream.Collectors#joining(CharSequence)`

Discussed [here](https://github.com/github/codeql/pull/11779#discussion_r1066677989).
 
`java.util.stream.Collectors#toMap(Function,Function)`

Similarly specialized collectors flow, not to be handled by MaD.

michaelnebel commented:
> > @atorralba : What is your view on simple types as sanitizers?
> 
> I agree that normally primitive types aren't likely to carry relevant taint. Nonetheless, there're a few queries that are interested in tainted primitives (e.g. a DoS by sleep, in which a tainted integer reaches a sleep function — an attacker could specify a high value and cause the program to sleep for a long time, practically causing a DoS). That's why in the previous round of top JDK APIs modelling, I recommended @jcogs33 to consider models that propagated taint through numbers (ints, longs, doubles...). Booleans and chars are different, but there's _probably_ a query out there that could be interested in them as well.
> 
> So the approach in my mind was adding these models, and then queries not interested in primitive types (the majority of them) would need to add a primitive type sanitizer. Most injection queries already do. There were discussions with @aschackmull about generalizing this kind of sanitizer so that we could reuse it more easily.
> 
> Now, if the model generator is excluding this kind of models by default, then I'm not sure if we should add them manually. Ideally the criteria for both generating and manually creating models should be the same. We could either modify the generator, or change our current criteria: instead of adding default taint steps for primitives types, queries interested in them should add additional taint steps specifically in their configurations. This could also be generalized as reusable taint steps, probably.
> 
> I'm open to either approach, so I'd be happy to hear more opinions.

Aha, that is interesting - I was unaware of the this example.
The reason for excluding the primitive types from the model generation is because that yields lots of extra models that are typically disregards and basically *disabled* by most sanitizers.

To also align with @yo-h suggestion, I think we should consider primitive types as sanitizers when it comes to models for this PR (as this is also mostly the case when looking into the existing hand written models), but add a comment that this might be revised. We need to address this as a general topics in the Models as Data area and I would very much like @aschackmull input as well.

jcogs33 commented:
Thanks all! I've switched the numeric flow models to neutrals in https://github.com/github/codeql/pull/11779/commits/6bb865ad0511c7114453b8d076135f9e85fdbcbf. Let me know if you want any changes to how I stored/commented them.

> We'll then probably need to do the same for numeric flow models added in the previous PR, for consistency.

I updated all of the numeric flow models from the previous PR as well (`java.lang.Integer#parseInt(String)`, `java.lang.String#valueOf(int)`, `java.math.BigDecimal#BigDecimal(String)`).

>  java.util.function.Consumer#accept(Object)
This is related to specialized lambda flow, so I don't think it should be handled by MaD.

@atorralba Should `java.util.function.Function#apply(Object)` not be handled by MaD either? I had added a model for it but am questioning whether it should be added since it is also related to lambda flow.

atorralba commented:
> @atorralba Should `java.util.function.Function#apply(Object)` not be handled by MaD either? I had added a model for it but am now questioning whether it should be added since it is also related to lambda flow.

Yes, without knowledge of what the function does, assuming that flow is propagated from arg0 to return value is too "greedy" and prone to FP, I think. Doing that with precision requires analyzing the code of the function itself, and building the dataflow graph accordingly (which, if I'm not mistaken, the existing lambda flow support already does, at least in some cases).

michaelnebel commented on java/ql/lib/ext/java.util.model.yml:
```diff
@@ -360,21 +362,34 @@ extensions:
+      - ["java.util", "Collections", "emptyMap", "()", "manual"]
+      - ["java.util", "Collections", "emptySet", "()", "manual"]
       - ["java.util", "Iterator", "hasNext", "()", "manual"]
+      - ["java.util", "List", "clear", "()", "manual"]
```
We probably need a comment here stating that this neutral model is introduced since `WithoutElement` has not yet been implemented for Java.
When `WithoutElement` gets implemented, we should get a summary like
`List.clear : Argument[-1].WithoutElement -> Argument[-1]`

michaelnebel replied:
Is that something you need in general for Java? I can give it a go, if you think you can use it.

atorralba replied:
I was just curious. I'm certain it's a generally useful feature, but I don't have any specific cases in mind right now.

michaelnebel commented on java/ql/lib/ext/java.util.model.yml:
```diff
@@ -360,21 +362,34 @@ extensions:
       - ["java.util", "List", "contains", "(Object)", "manual"]
       - ["java.util", "List", "isEmpty", "()", "manual"]
       - ["java.util", "List", "size", "()", "manual"]
+      - ["java.util", "Map", "clear", "()", "manual"]
```
Most likely something similar as for List.clear.

jcogs33 replied:
Updated both in https://github.com/github/codeql/pull/11779/commits/fb6725ddaa4a62466b23046c36a686cee7e00da4.

michaelnebel reviewed (approved):
Looks good to me!
Minor: Added a comment about `List.clear` and `Map.clear`.

michaelnebel reviewed (approved):
LGTM!

atorralba commented on java/ql/lib/ext/java.lang.model.yml:
```diff
@@ -87,34 +89,60 @@ extensions:
       - ["java.lang", "StringBuffer", True, "StringBuffer", "(String)", "", "Argument[0]", "Argument[-1]", "taint", "manual"]
       - ["java.lang", "StringBuilder", True, "StringBuilder", "", "", "Argument[0]", "Argument[-1]", "taint", "manual"]
       - ["java.lang", "System", False, "arraycopy", "", "", "Argument[0]", "Argument[2]", "taint", "manual"]
+      - ["java.lang", "System", False, "getProperty", "(String)", "", "SyntheticGlobal[java.lang.System.properties].MapValue", "ReturnValue", "value", "manual"]
+      - ["java.lang", "System", False, "setProperty", "(String,String)", "", "SyntheticGlobal[java.lang.System.properties].MapValue", "ReturnValue", "value", "manual"]
+      - ["java.lang", "System", False, "setProperty", "(String,String)", "", "Argument[0]", "SyntheticGlobal[java.lang.System.properties].MapKey", "value", "manual"]
+      - ["java.lang", "System", False, "setProperty", "(String,String)", "", "Argument[1]", "SyntheticGlobal[java.lang.System.properties].MapValue", "value", "manual"]
```
After reviewing the DCA results, it seems that these 4 need to be removed. They cause way too much spurious flow in projects that make extensive use of `System.[get|set]Property`. These ought to be modeled by regular CodeQL matching the get and set keys to reduce FPs (not in this PR).

jcogs33 replied:
Removed in https://github.com/github/codeql/pull/11779/commits/10f0975812bc6c3faf0b8a22f6bdb5331aa7e780.

atorralba reviewed (approved):
Thanks for all the effort! 🙏
